src/main/java/edu/harvard/iq/dataverse/search/IndexServiceBean.java

```diff
@@ -1405,17 +1435,29 @@ private String addOrUpdateDataset(IndexableDataset indexableDataset, Set<Long> d
                 throw new IOException(ex);
             }
         }
+        return docs.getMessage();
+    }
+
+    @Asynchronous
```

Comment:
At least this is the goal of isolating the update operation in the smallest possible transaction. The only locking problems I still have are when ingesting needs to make thumbnails, it takes an ingest lock on the dataset that prevents editting.

Also, if the locking exception would happen on the transaction here, it would be OK since it means that the dataset is being updated and it will be reindexed anyway.